dir/fs.go

```diff
@@ -0,0 +1,128 @@
+package dir
+
+import (
+	"errors"
+	"io/fs"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+// UnionDirFsIFace is a simple union directory file system interface
+type UnionDirFsIFace interface {
+	fs.ReadDirFS
+	GetPath(elem ...string) (string, error)
+}
+
+// NewUnionDirFS is a virtual file system for merging multiple directories with priority
+//
+// dirs contains the union directories and the previous one has higher priority
+func NewUnionDirFS(dirs ...string) UnionDirFS {
+	rootedDirs := []RootedFS{}
+	for _, dir := range dirs {
+		rootedDirs = append(rootedDirs, RootedFS{os.DirFS(dir), dir})
+	}
+	return UnionDirFS{Dirs: rootedDirs}
+}
+
+// RootedFS is io.FS implementation used in New.
+// root is the root of the file system tree passed to os.DirFS.
+type RootedFS struct {
+	fs.FS
+	Root string
+}
+
+// UnionDirFS is a simple union directory file system
+//
+// it unions multiple directory to be one directory with priority based on the order of Dirs
+type UnionDirFS struct {
+	Dirs []RootedFS
+}
+
+// Open implements fs.FS interface
+//
+// traverse all union directories and return the first existing file
+func (u UnionDirFS) Open(name string) (fs.File, error) {
+	for _, dir := range u.Dirs {
+		f, err := dir.Open(name)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return nil, err
+			}
+		}
+		return f, nil
+	}
+	return nil, fs.ErrNotExist
+}
+
+// Path returns the path of the named file or directory under the FS
+//
+// if path exists, it returns the first existing path in union directories (dirs)
+// if path doesn't exist, it returns fs.ErrNotExist error
+func (u UnionDirFS) GetPath(elem ...string) (string, error) {
+	pathSuffix := path.Join(elem...)
+	for _, dir := range u.Dirs {
+		file, err := dir.Open(pathSuffix)
+		if err != nil {
+			if errors.Is(err, fs.ErrNotExist) {
+				continue
+			} else {
+				return "", err
+			}
+		}
+		defer file.Close()
+		// got the first existing file path and break
+		// return the path with current OS separator
+		return filepath.Join(dir.Root, pathSuffix), nil
+	}
+	return "", fs.ErrNotExist
+}
+
+// ReadDir implements the ReadDirFile interface
+//
+// traverse all union directories and return all existing DirEntries
+func (u UnionDirFS) ReadDir(name string) ([]fs.DirEntry, error) {
+	isVisited := make(map[string]bool)
```

Comment:
I'm stilling catching up on golang, but looks like you are trying to create a Set data structure here but instead using a Map. If it is the convention for golang, should map[T]struct{} a cheaper option than map[string]bool in terms of number of bytes used? After all, you are just checking if T is in the Map or not.

Reply:
Sure. Done.